# Project Euler The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

In [1]:
buffer = ["73167176531330624919225119674426574742355349194934",
          "96983520312774506326239578318016984801869478851843",
          "85861560789112949495459501737958331952853208805511",
          "12540698747158523863050715693290963295227443043557",
          "66896648950445244523161731856403098711121722383113",
          "62229893423380308135336276614282806444486645238749",
          "30358907296290491560440772390713810515859307960866",
          "70172427121883998797908792274921901699720888093776",
          "65727333001053367881220235421809751254540594752243",
          "52584907711670556013604839586446706324415722155397",
          "53697817977846174064955149290862569321978468622482",
          "83972241375657056057490261407972968652414535100474",
          "82166370484403199890008895243450658541227588666881",
          "16427171479924442928230863465674813919123162824586",
          "17866458359124566529476545682848912883142607690042",
          "24219022671055626321111109370544217506941658960408",
          "07198403850962455444362981230987879927244284909188",
          "84580156166097919133875499200524063689912560717606",
          "05886116467109405077541002256983155200055935729725",
          "71636269561882670428252483600823257530420752963450"]

print(buffer)

['73167176531330624919225119674426574742355349194934', '96983520312774506326239578318016984801869478851843', '85861560789112949495459501737958331952853208805511', '12540698747158523863050715693290963295227443043557', '66896648950445244523161731856403098711121722383113', '62229893423380308135336276614282806444486645238749', '30358907296290491560440772390713810515859307960866', '70172427121883998797908792274921901699720888093776', '65727333001053367881220235421809751254540594752243', '52584907711670556013604839586446706324415722155397', '53697817977846174064955149290862569321978468622482', '83972241375657056057490261407972968652414535100474', '82166370484403199890008895243450658541227588666881', '16427171479924442928230863465674813919123162824586', '17866458359124566529476545682848912883142607690042', '24219022671055626321111109370544217506941658960408', '07198403850962455444362981230987879927244284909188', '84580156166097919133875499200524063689912560717606', '05886116467109405077541002

In [2]:
print(buffer[1])

96983520312774506326239578318016984801869478851843


In [3]:
print(buffer[1][0])

9


In [4]:
print(len(buffer[0]))
print(len(buffer))

50
20


In [5]:
def prod_list(numbers):
    #print(numbers)
    if numbers.count(0) >= 1:
        return 0
    prod = 1
    for number in numbers:
        prod *= number
    return prod

In [6]:
import numpy
import random

print("Testing prod_list function...")

list = []
numpy_error = 0
my_error = 0
real_prod = 1
for i in range(0,10000):
    if (i % 1000) == 0:
        list.clear()
        real_prod = 1
    digit = random.randint(1,9999)
    list.append(digit)
    real_prod *= digit
    # cross checking function with numpy
    if prod_list(list) != numpy.prod(list):
        numpy_error += 1
        #print(list)
        #print(prod_list(list))
        #print(numpy.prod(list))
    if prod_list(list) != real_prod:
        my_error += 1
print("errors in numpy.prod: ", numpy_error/10000)
print("errors in prod_list: ", my_error/10000)

Testing prod_list function...
errors in numpy.prod:  0.9979
errors in prod_list:  0.0


For some reason numpy.prod does not seem reliable 99.8% of the time. Looking closer sometimes it returns negative numbers likely as a result of overflow (only use positive numbers) so creating a custom function ad unit testing was probably the best way forward with this problem.

Also it tests at varied ranges of lists (1 to 1000) and larger variation of numbers than single digits (up to 4 digits).
This should instill a high confidence in the prod_list() function at producing correct products. 

In [7]:
import time
start_time = time.time()

# just directly fill in the first 13 numbers
list = [7,3,1,6,7,1,7,6,5,3,1,3,3]

i = 0
j = 13
max_prod_wrap = prod_list(list)
max_prod_no_wrap = prod_list(list)
while True:
    list.pop(0)
    list.append(int(buffer[i][j]))  # need numeric value
    
    curr_prod = prod_list(list)
    if curr_prod > max_prod_wrap:
        max_prod_wrap = curr_prod
    if (j >= 12) & (curr_prod > max_prod_no_wrap):
        max_prod_no_wrap = curr_prod
    
    # iterate i and j in 20 by 50 char matrix
    j += 1
    if j >= 50:
        j = 0
        i += 1
        
    if i >= 20:
        break 

print("max prod with wrap:     ", max_prod_wrap)
print("max prod without wrap:  ", max_prod_no_wrap)

print("--- %s seconds ---" % (time.time() - start_time))

max prod with wrap:      23514624000
max prod without wrap:   5377010688
--- 0.003964900970458984 seconds ---


Couldn't run any solution online to compare, but it seems like the correct.
23,514,624,000 for wrapping text which doesnt seem to fit the question
5,377,010,688 for no wrapping text which does seem to fit the question